File: javascript/src/llm/runner.ts
/**
 * SKILL.md steps 1–2 (my-turn + move) and optional agentBasics for steps 3–4 starter subset.
 * @see https://moltchess.com/skill.md — route details: https://moltchess.com/api-docs/llms.txt
 */
import { MoltChessApiError, type MoltChessClient } from "../client.js";
import { type AgentBasicsConfig, runAgentBasicsOnce } from "./agentBasics.js";
import { firstLegalFallback, legalMovesFromFen } from "./board.js";
import type { LlmMoveChooser, MovePromptContext } from "./types.js";
import { resolveLegalMove } from "./validate.js";

function coerceMyTurnGames(raw: unknown): Record<string, unknown>[] {
  if (Array.isArray(raw)) {
    return raw.filter((g): g is Record<string, unknown> => typeof g === "object" && g !== null);
  }
  if (raw && typeof raw === "object") {
    const obj = raw as Record<string, unknown>;
    for (const key of ["games", "data", "items", "results"] as const) {
      const inner = obj[key];
      if (Array.isArray(inner)) {
        return inner.filter((g): g is Record<string, unknown> => typeof g === "object" && g !== null);
      }
    }
  }
  return [];
}

function gameIdFromSummary(row: Record<string, unknown>): string | null {
  for (const key of ["game_id", "id", "uuid"] as const) {
    const val = row[key];
    if (val !== undefined && val !== null && String(val).length > 0) {
      return String(val);
    }
  }
  return null;
}

function movesSummary(moves: unknown): string {
  if (!Array.isArray(moves)) {
    return "";
  }
  const parts: string[] = [];
  for (const m of moves) {
    if (m && typeof m === "object") {
      const san = (m as Record<string, unknown>).move_san;
      if (typeof san === "string" && san.length > 0) {
        parts.push(san);
      }
    }
  }
  return parts.join(" ");
}

export type LogFn = (msg: string) => void;

function errorMessage(error: unknown): string {
  if (error instanceof Error) {
    return error.message;
  }
  return String(error);
}

async function chooseMoveWithRetry(
  chooser: LlmMoveChooser,
  ctx: MovePromptContext,
  legalUci: string[],
  uciToSan: Map<string, string>,
  log?: LogFn | null,
): Promise<{ san: string; uci: string } | null> {
  let feedback: string | null = null;
  for (let attempt = 1; attempt <= 2; attempt += 1) {
    try {
      const choice = await chooser.chooseMove(ctx, feedback);
      const resolved = resolveLegalMove(ctx.fen, choice, legalUci, uciToSan);
      if (resolved) {
        return resolved;
      }
      feedback = "Your JSON did not contain a legal move_san or move_uci from the lists.";
    } catch (error) {
      feedback = `The previous call failed: ${errorMessage(error)}`;
      log?.(`[llm] chooser error for ${ctx.gameId} (attempt ${attempt}): ${feedback}`);
    }
  }
  return null;
}

export async function runLlmHeartbeatOnce(
  client: MoltChessClient,
  chooser: LlmMoveChooser,
  options: {
    myTurnLimit?: number;
    log?: LogFn | null;
    agentBasics?: AgentBasicsConfig | null;
  } = {},
): Promise<string[]> {
  const myTurnLimit = options.myTurnLimit ?? 50;
  const log = options.log ?? console.log;
  const lines: string[] = [];
  const raw = await client.chess.getMyTurnGames({ limit: myTurnLimit });
  const summaries = coerceMyTurnGames(raw);

  for (const summary of summaries) {
    const gid = gameIdFromSummary(summary);
    if (!gid) {
      continue;
    }
    let detail: unknown;
    try {
      detail = await client.chess.getGame(gid);
    } catch (err) {
      if (err instanceof MoltChessApiError) {
        log?.(`[llm] skip game (fetch failed): ${gid} ${err.message}`);
      }
      continue;
    }
    if (!detail || typeof detail !== "object") {
      continue;
    }
    const d = detail as Record<string, unknown>;
    if (!d.is_my_turn) {
      continue;
    }
    const fen = d.current_fen;
    if (typeof fen !== "string" || !fen.trim()) {
      log?.(`[llm] skip game (no FEN): ${gid}`);
      continue;
    }
    const myColor = String(d.my_color ?? "white");
    const summaryText = movesSummary(d.moves);

    const { legalSans, legalUci, uciToSan } = legalMovesFromFen(fen);
    if (!legalUci.length) {
      log?.(`[llm] skip game (no legal moves): ${gid}`);
      continue;
    }

    const ctx: MovePromptContext = {
      gameId: gid,
      fen,
      myColor,
      legalSans,
      legalUci,
      movesSummary: summaryText,
    };

    let resolved = await chooseMoveWithRetry(chooser, ctx, legalUci, uciToSan, log);
    if (!resolved) {
      const fb = firstLegalFallback({ legalSans, legalUci, uciToSan });
      log?.(`[llm] fallback move for ${gid}: ${fb.san} (${fb.uci})`);
      resolved = fb;
    }

    try {
      await client.chess.submitMove({
        game_id: gid,
        move_san: resolved.san,
        move_uci: resolved.uci,
      });
    } catch (err) {
      const msg = err instanceof MoltChessApiError ? err.message : String(err);
      const line = `[llm] submit_move failed ${gid}: ${msg}`;
      lines.push(line);
      log?.(line);
      continue;
    }

    const line = `[llm] played ${resolved.san} (${resolved.uci}) in ${gid}`;
    lines.push(line);
    log?.(line);
  }

  if (options.agentBasics !== undefined && options.agentBasics !== null) {
    const extra = await runAgentBasicsOnce(client, options.agentBasics, { log });
    lines.push(...extra);
  }

  return lines;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export async function runLlmHeartbeatLoop(
  client: MoltChessClient,
  chooser: LlmMoveChooser,
  options: {
    intervalSec?: number;
    myTurnLimit?: number;
    log?: LogFn | null;
    stopAfterOne?: boolean;
    agentBasics?: AgentBasicsConfig | null;
  } = {},
): Promise<void> {
  const intervalSec = options.intervalSec ?? 45;
  const myTurnLimit = options.myTurnLimit ?? 50;
  const log = options.log ?? console.log;
  const stopAfterOne = options.stopAfterOne ?? false;

  for (;;) {
    try {
      await runLlmHeartbeatOnce(client, chooser, {
        myTurnLimit,
        log,
        agentBasics: options.agentBasics,
      });
    } catch (error) {
      log?.(`[llm] heartbeat tick failed: ${errorMessage(error)}`);
    }
    if (stopAfterOne) {
      return;
    }
    await sleep(intervalSec * 1000);
  }
}

File: python/src/moltchess/llm/runner.py
"""LLM heartbeat: SKILL.md steps 1–2 (my-turn games, then POST /api/chess/move) plus optional agent_basics."""

from __future__ import annotations

import time
from typing import Any, Callable, Mapping

from ..client import MoltChessApiError, MoltChessClient
from .agent_basics import AgentBasicsConfig, run_agent_basics_once
from .board import first_legal_fallback, legal_moves_from_fen
from .types import LlmMoveChooser, MovePromptContext
from .validate import resolve_legal_move

LogFn = Callable[[str], None]


def _error_message(error: object) -> str:
    if isinstance(error, Exception):
        return str(error)
    return repr(error)


def _coerce_my_turn_games(raw: Any) -> list[Mapping[str, Any]]:
    if isinstance(raw, list):
        return [g for g in raw if isinstance(g, Mapping)]
    if isinstance(raw, Mapping):
        for key in ("games", "data", "items", "results"):
            inner = raw.get(key)
            if isinstance(inner, list):
                return [g for g in inner if isinstance(g, Mapping)]
    return []


def _game_id_from_summary(row: Mapping[str, Any]) -> str | None:
    for key in ("game_id", "id", "uuid"):
        val = row.get(key)
        if val:
            return str(val)
    return None


def _moves_summary(moves: Any) -> str:
    if not isinstance(moves, list):
        return ""
    parts: list[str] = []
    for m in moves:
        if isinstance(m, Mapping):
            san = m.get("move_san")
            if isinstance(san, str) and san:
                parts.append(san)
    return " ".join(parts)


def run_llm_heartbeat_once(
    client: MoltChessClient,
    chooser: LlmMoveChooser,
    *,
    my_turn_limit: int = 50,
    log: LogFn | None = print,
    agent_basics: AgentBasicsConfig | None = None,
) -> list[str]:
    lines: list[str] = []
    raw = client.chess.get_my_turn_games(limit=my_turn_limit)
    summaries = _coerce_my_turn_games(raw)

    for summary in summaries:
        gid = _game_id_from_summary(summary)
        if not gid:
            continue
        try:
            detail = client.chess.get_game(gid)
        except MoltChessApiError as exc:
            if log:
                log(f"[llm] skip game (fetch failed): {gid} {exc}")
            continue
        if not isinstance(detail, Mapping):
            continue
        if not detail.get("is_my_turn"):
            continue
        fen = detail.get("current_fen")
        if not isinstance(fen, str) or not fen.strip():
            if log:
                log(f"[llm] skip game (no FEN): {gid}")
            continue
        my_color = str(detail.get("my_color") or "white")
        moves = detail.get("moves")
        summary_text = _moves_summary(moves)

        legal_sans, legal_uci, uci_to_san = legal_moves_from_fen(fen)
        if not legal_uci:
            if log:
                log(f"[llm] skip game (no legal moves): {gid}")
            continue

        ctx = MovePromptContext(
            game_id=gid,
            fen=fen,
            my_color=my_color,
            legal_sans=legal_sans,
            legal_uci=legal_uci,
            moves_summary=summary_text,
        )

        resolved = None
        feedback: str | None = None
        for attempt in range(1, 3):
            try:
                choice = chooser.choose_move(ctx, feedback=feedback)
                resolved = resolve_legal_move(fen, choice, legal_sans, legal_uci, uci_to_san)
                if resolved is not None:
                    break
                feedback = "Your JSON did not contain a legal move_san or move_uci from the lists."
            except Exception as exc:  # noqa: BLE001 - providers may raise transport-specific exceptions
                feedback = f"The previous call failed: {_error_message(exc)}"
                if log:
                    log(f"[llm] chooser error for {gid} (attempt {attempt}): {feedback}")
        if resolved is None:
            san_fb, uci_fb = first_legal_fallback(legal_sans, legal_uci, uci_to_san)
            if log:
                log(f"[llm] fallback move for {gid}: {san_fb} ({uci_fb})")
            resolved = (san_fb, uci_fb)

        move_san, move_uci = resolved
        try:
            client.chess.submit_move(
                {
                    "game_id": gid,
                    "move_san": move_san,
                    "move_uci": move_uci,
                }
            )
        except MoltChessApiError as exc:
            msg = f"[llm] submit_move failed {gid}: {exc}"
            lines.append(msg)
            if log:
                log(msg)
            continue

        msg = f"[llm] played {move_san} ({move_uci}) in {gid}"
        lines.append(msg)
        if log:
            log(msg)

    if agent_basics is not None:
        lines.extend(run_agent_basics_once(client, agent_basics, log=log))

    return lines


def run_llm_heartbeat_loop(
    client: MoltChessClient,
    chooser: LlmMoveChooser,
    *,
    interval_sec: float = 45.0,
    my_turn_limit: int = 50,
    log: LogFn | None = print,
    stop_after_one: bool = False,
    agent_basics: AgentBasicsConfig | None = None,
) -> None:
    while True:
        try:
            run_llm_heartbeat_once(
                client,
                chooser,
                my_turn_limit=my_turn_limit,
                log=log,
                agent_basics=agent_basics,
            )
        except Exception as exc:  # noqa: BLE001 - keep long-running loops alive
            if log:
                log(f"[llm] heartbeat tick failed: {_error_message(exc)}")
        if stop_after_one:
            return
        time.sleep(interval_sec)
